File: oc-xcode插件/xcodeSourceEditorLeaning/Snowonder Extension/SourceEditorCommand.m
//
//  SourceEditorCommand.m
//  Snowonder Extension
//
//  Created by Aron Ruan on 2022/5/5.
//

#import "SourceEditorCommand.h"

//extension String {
//
//    func matches(pattern: String) -> Bool {
//        guard let regularExpression = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else {
//            return false
//        }
//
//        let matchingStrings = regularExpression.matches(in: self, range: NSMakeRange(0, count))
//        return !matchingStrings.isEmpty
//    }
//}

@implementation SourceEditorCommand

- (void)performCommandWithInvocation:(XCSourceEditorCommandInvocation *)invocation completionHandler:(void (^)(NSError * _Nullable nilOrError))completionHandler
{
    // Implement your command here, invoking the completion handler when done. Pass it nil on success, and an NSError on failure.
    
    NSMutableArray<NSString *> * lines = invocation.buffer.lines;
    NSLog(@"lines = %@",lines);
    
    NSError *error = nil;
    
    if([lines isKindOfClass:[NSMutableArray<NSString*> class]])
    {
        @try {
            ImportBlockDetector * detector=[[ImportBlockDetector alloc]init];
            
            
            
        } @catch (NSError *exception) {
//            <#Handle an exception thrown in the @try block#>
            error = exception;
        } @finally {
            
        }
    }
    
    completionHandler(error);
}

@end

@implementation ImportBlock

@end

@implementation ImportBlockDetector

//open func importBlock(from lines: [String]) throws -> ImportBlock {
//    let group = self.group(for: lines, using: Constant.allGroups)
//
//    guard !group.isEmpty else {
//        throw ImportBlockDetectorError.notFound
//    }
//
//    let declarations = self.declarations(from: lines, using: group)
//    let categorizedDeclarations = self.categorizedDeclarations(from: declarations, using: group)
//
//    return .init(group: group, declarations: declarations, categorizedDeclarations: categorizedDeclarations)
//}
-(ImportBlock*)importBlock:(NSMutableArray<NSString*>*)lines
{
    ImportBlock * s= [[ImportBlock alloc] init];
    
    return s;
}

//open func group(for lines: [String], using availableGroups: [ImportCategoriesGroup]) -> ImportCategoriesGroup {
//    for line in lines {
//        for group in availableGroups {
//            for category in group {
//                if line.matches(pattern: category.declarationPattern) {
//                    return group
//                }
//            }
//        }
//    }
//
//    return .init()
//}
-(NSMutableArray<ImportCategory*>*)group:(NSMutableArray<NSString*>*)lines :(NSMutableArray<NSMutableArray<ImportCategory*>*>*)availableGroups
{
    NSMutableArray<ImportCategory*>* s =[[NSMutableArray alloc]init];
//    for(NSString *line in lines)
//    {
//        for(NSMutableArray<ImportCategory*>* group in availableGroups)
//        {
//            for(ImportCategory * category in group)
//            {
//                
//            }
//        }
//    }
    return s;
}

@end

@implementation ImportCategory

@end

File: oc-xcode插件/Snowonder-swift/Snowonder Extension swift/SourceEditorCommand.swift
//
//  SourceEditorCommand.swift
//  Snowonder Extension swift
//
//  Created by Aron Ruan on 2022/5/5.
//

import Foundation
import XcodeKit

class SourceEditorCommand: NSObject, XCSourceEditorCommand {
    
    func perform(with invocation: XCSourceEditorCommandInvocation, completionHandler: @escaping (Error?) -> Void ) -> Void {
        // Implement your command here, invoking the completion handler when done. Pass it nil on success, and an NSError on failure.
        
        var error: Error? = nil
        
        if let lines = invocation.buffer.lines as? [String] {
            do {
                let detector = ImportBlockDetector()
                let importBlock = try detector.importBlock(from: lines)
                let formatter = ImportBlockFormatter()
                let formattedImportLines = formatter.lines(for: importBlock, using: ImportBlockFormatter.Operation.all)

                let bufferEditor = SourceTextBufferEditor(buffer: invocation.buffer)
                bufferEditor.replace(lines: importBlock.declarations, with: formattedImportLines, using: .top)
            } catch let catchedError {
                error = catchedError
            }
        }
        
        completionHandler(error)
    }
    
}
